packages/middleware/rate-limiter/src/implementation/redis.ts

```diff
@@ -0,0 +1,78 @@
+import { Slot, SlotProvider } from '../slot'
+import { RedisClientType } from '@redis/client'
+
+class RedisSlot implements Slot {
+  readonly fromSecond: number
+  readonly durationSecond: number
+  readonly key: string
+  readonly client: RedisClientType
+  private counter: number
+
+  constructor({
+    fromSecond,
+    durationSecond,
+    key,
+    client,
+  }: {
+    fromSecond: number
+    durationSecond: number
+    key: string
+    client: RedisClientType
+  }) {
+    this.counter = 0
+    this.fromSecond = fromSecond
+    this.durationSecond = durationSecond
+    this.key = `${key}:${fromSecond}`
+    this.client = client
+  }
+
+  inc(): void {
+    this.client
+      .incr(this.key)
+      .catch(() => console.error('REDIS: Failed to incr', this.key))
+      .then((value) => {
+        if (this.counter === 0) {
+          this.client
+            .expireAt(this.key, new Date((this.fromSecond + this.durationSecond + 60) * 1000), 'NX')
+            .catch(() => console.error('REDIS: Failed to set expiration', this.key))
+            .then(() => {})
+        }
+        if (value != null) {
+          this.counter = value
+        }
+      })
+  }
+
+  value(): number {
+    this.client
+      .get(this.key)
+      .catch(() => console.error('REDIS: Failed to get', this.key))
+      .then((value) => {
+        if (value != null) {
+          const v = Number(value)
+          if (!Number.isNaN) {
+            this.counter = v
+          }
+        }
+      })
```

Comment:
No need to be over. We don't want to slow down the request only to be 100% sure the user can call again the service. One or two requests that should be rate-limited but still pass is not a big problem.

Reply:
Thanks for taking the time to explain this to me!
Maybe you could add it to a comment explaining this, to avoid that future maintainers will be puzzled by this since it's not immediately clear from the code